circuits/src/zk_circuits/valid_offline_fee_settlement.rs

```diff
@@ -0,0 +1,356 @@
+            &witness.original_wallet_public_shares,
+            &witness.original_wallet_private_shares,
+            &witness.merkle_opening,
+            statement.merkle_root,
+            statement.nullifier,
+            cs,
+        )?;
+
+        // Check the commitment to the new wallet shares
+        let new_wallet_comm =
+            WalletGadget::compute_private_commitment(&witness.updated_wallet_private_shares, cs)?;
+        EqGadget::constrain_eq(&new_wallet_comm, &statement.updated_wallet_commitment, cs)?;
+
+        // Reconstruct the old and new wallets
+        let old_wallet = WalletGadget::wallet_from_shares(
+            &witness.original_wallet_public_shares,
+            &witness.original_wallet_private_shares,
+            cs,
+        )?;
+        let new_wallet = WalletGadget::wallet_from_shares(
+            &statement.updated_wallet_public_shares,
+            &witness.updated_wallet_private_shares,
+            cs,
+        )?;
+
+        // --- Note Construction & Encryption --- //
+        // Select the encryption key to use based on whether the fee is being paid to
+        // the protocol
+        let key = CondSelectGadget::select(
+            &statement.protocol_key,
+            &old_wallet.managing_cluster,
+            statement.is_protocol_fee,
+            cs,
+        )?;
+
+        // Verify that the note correctly spends the balance
+        let send_bal = Self::get_balance_at_idx(witness.send_index, &old_wallet, cs)?;
+        Self::verify_note_construction(
+            &witness.note,
+            statement.is_protocol_fee,
+            &send_bal,
+            &key,
+            cs,
+        )?;
+
+        // Verify the note encryption and its commitment
+        Self::verify_note_encryption_commitment(
+            &witness.note,
+            &key,
+            witness.encryption_randomness,
+            &statement.note_ciphertext,
+            statement.note_commitment,
+            cs,
+        )?;
+
+        // --- Reblind and State Transition --- //
+
+        // Verify that the sender's wallet is correctly reblinded and updated
+        Self::verify_sender_reblind(
+            &witness.original_wallet_private_shares,
+            &witness.updated_wallet_private_shares,
+            &statement.updated_wallet_public_shares,
+            cs,
+        )?;
+        Self::verify_sender_state_transition(
+            witness.send_index,
+            statement.is_protocol_fee,
+            &old_wallet,
+            &new_wallet,
+            cs,
+        )
+    }
+
+    /// Verify coherence between the note and the send balance
+    fn verify_note_construction(
+        note: &NoteVar,
+        is_protocol_fee: BoolVar,
+        send_bal: &BalanceVar,
+        enc_key: &EncryptionKeyVar,
+        cs: &mut PlonkCircuit,
+    ) -> Result<(), CircuitError> {
+        // The note's mint should match the send balance
+        EqGadget::constrain_eq(&note.mint, &send_bal.mint, cs)?;
```

Comment:
do we also need the check that $\text{Note}.\text{mint} \neq 0$?

Reply:
fixed in follow up